azurelinuxagent/common/osutil/default.py

```diff
@@ -1203,25 +1203,43 @@ def device_for_ide_port(self, port_id):
             port_id = port_id - 2
         device = None
         path = "/sys/bus/vmbus/devices/"
-        if os.path.exists(path):
-            try:
-                for vmbus in os.listdir(path):
-                    deviceid = fileutil.read_file(os.path.join(path, vmbus, "device_id"))
-                    guid = deviceid.lstrip('{').split('-')
-                    if guid[0] == g0 and guid[1] == "000" + ustr(port_id):
-                        for root, dirs, files in os.walk(path + vmbus):
-                            if root.endswith("/block"):
-                                device = dirs[0]
-                                break
-                            else:
-                                # older distros
-                                for d in dirs:
-                                    if ':' in d and "block" == d.split(':')[0]:
-                                        device = d.split(':')[1]
-                                        break
+        # We have to try device IDs for both Gen1 and Gen2 VMs.
+        gen2_device_id = 'f8b3781a-1e82-4818-a1c3-63d806ec15bb'
+        for device_id_or_prefix in [
+                '{0}-000{1}'.format(g0, port_id),
+                gen2_device_id]:
+            logger.info('Looking for device ID with name or prefix {0}'.format(device_id_or_prefix))
+            if os.path.exists(path):
+                try:
+                    for vmbus in os.listdir(path):
+                        deviceid = fileutil.read_file(os.path.join(path, vmbus, "device_id"))
+                        guid = deviceid.strip('{}')
+                        if guid.startswith(device_id_or_prefix):
+                            for root, dirs, files in os.walk(path + vmbus):
+                                root_path_parts = root.split('/')
+                                # For Gen1 VMs we only have to check for the block dir in the
+                                # current device. But for Gen2 VMs all of the disks (sda, sdb,
+                                # sr0) are presented in this device on the same SCSI controller.
+                                # Because of that we need to also read the LUN. It will be:
+                                #   0 - OS disk
+                                #   1 - Resource disk
+                                #   2 - CDROM
+                                if root_path_parts[-1] == 'block' and (
```

Comment:
Can you please add a unit test to mock the success and failure conditions, such as this code is covered?